modules/gateway/nodes.go

```diff
@@ -90,68 +97,134 @@ func (g *Gateway) requestNodes(conn modules.PeerConn) error {
-func (g *Gateway) threadedNodeManager() {
-	if g.threads.Add() != nil {
-		return
-	}
-	defer g.threads.Done()
+// permanentNodePurger is a thread that runs throughout the lifetime of the
+// gateway, purging unconnectable nodes from the node list in a sustainable
+// way.
+func (g *Gateway) permanentNodePurger(closeChan chan struct{}) {
+	defer close(closeChan)
 
 	for {
+		// Start by sleeping for 10 minutes. This means that no nodes will be
+		// purged for the first 10 minutes that Sia is running, and that any
+		// failures to get nodes will be met by 10 minutes of waiting.
+		//
+		// At most one node will be contacted every 10 minutes. This minimizes
+		// the total amount of keepalive traffic on the network.
 		select {
-		case <-time.After(5 * time.Second):
+		case <-time.After(nodePurgeDelay):
 		case <-g.threads.StopChan():
+			// The gateway is shutting down, close out the thread.
 			return
 		}
 
+		// Get a random node for scanning.
 		g.mu.RLock()
 		numNodes := len(g.nodes)
-		peer, err := g.randomPeer()
+		node, err := g.randomNode()
 		g.mu.RUnlock()
 		if err != nil {
-			// can't do much until we have peers
+			// Most likely the error indicates that there are no nodes in the
+			// node list. Wait until the next iteration to try again.
+			continue
+		}
+		if numNodes <= pruneNodeListLen {
+			// There are not enough nodes in the gateway - pruning more is
+			// probably a bad idea, and may affect the user's ability to
+			// connect to the network in the future.
 			continue
 		}
 
-		if numNodes < minNodeListLen {
-			err := g.RPC(peer, "ShareNodes", g.requestNodes)
+		// Try connecting to the random node. If the node is not reachable,
+		// remove them from the node list. Make sure that the dial is stopped
+		// early if the gateway is shutting down.
+		func() {
+			cancelDialChan := make(chan struct{})
+			dialDoneChan := make(chan struct{})
+			defer close(dialDoneChan)
+			go func() {
+				select {
+				case <-dialDoneChan:
+				case <-g.threads.StopChan():
+					close(cancelDialChan)
+				}
+			}()
+			dialer := &net.Dialer{
+				Timeout: dialTimeout,
+				Cancel:  cancelDialChan,
+			}
+			conn, err := dialer.Dial("tcp", string(node))
 			if err != nil {
-				g.log.Debugf("WARN: RPC ShareNodes failed on peer %q: %v", peer, err)
-				continue
+				// NOTE: an error may be returned if the dial is cancelled
+				// partway through, which would cause the node to be pruned
+				// even though it may be a good node. Because nodes are
+				// plentiful, that's not a huge problem.
+				g.mu.Lock()
+				g.removeNode(node)
+				g.save()
+				g.mu.Unlock()
+				g.log.Debugf("INFO: removing node %q because dialing it failed: %v", node, err)
+				return
 			}
+
+			// If connection succeeds, supply an unacceptable version so that we
+			// will not be added as a peer.
+			//
+			// NOTE: this is a somewhat clunky way of specifying that you didn't
+			// actually want a connection.
+			encoding.WriteObject(conn, "0.0.0")
```

Comment:
it's also annoying because it results in lots of `"WARN: peer tried to connect, but their version (0.0.0) was unacceptable"` log lines. We'll probably want to support this "ping" behavior next time we change the gateway protocol.